### PR TITLE
Backend server authentication. Store name, message and picture

### PR DESCRIPTION
**Pull request by swalsh15**
Sends a user ID token to backend when user submits comment. Uses GoogleIdTokenVerifier to authenticate the user and then accesses the user's name and picture url. Stores the comment, name, and picture in datastore. Send the post request from javascript instead of the html form.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -69,16 +76,30 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-    if (comment == null || name == null) {
+    GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(new NetHttpTransport(), JacksonFactory.getDefaultInstance())
+      .setAudience(Collections.singletonList("813751014340-ht5ugfu1pqj5a4gqq7rbvjk0sffu00it.apps.googleusercontent.com")).build();
+    GoogleIdToken idToken = null;     
+    String key = request.getParameter("idtoken");
+    
+    try {
+      idToken = verifier.verify(key);
+    } catch (Exception e) {
+      return;
```
Comment (g-harel) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:87]: Would be good to return an auth error with a status code in the 400 range.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -69,16 +76,30 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-    String name = request.getParameter("name_input");
-    if (comment == null || name == null) {
+    GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(new NetHttpTransport(), JacksonFactory.getDefaultInstance())
+      .setAudience(Collections.singletonList("813751014340-ht5ugfu1pqj5a4gqq7rbvjk0sffu00it.apps.googleusercontent.com")).build();
+    GoogleIdToken idToken = null;     
+    String key = request.getParameter("idtoken");
+    
+    try {
+      idToken = verifier.verify(key);
+    } catch (Exception e) {
```
Comment (g-harel) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:86]: Does this throw if verification fails?

**Comment by g-harel** [comment]
Very cool so far!

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -71,30 +71,40 @@ function fetchComments() {
...
     document.getElementById("comment-form").style.display = "none";
   }
-}
\ No newline at end of file
+}
+
+function commentPostRequest() {
+  const id_token = gapi.auth2.getAuthInstance().currentUser.get().getAuthResponse().id_token;
+  const textbox = document.getElementById("text_input");
+  const xhr = new XMLHttpRequest();
```
Comment (augerm) [inline portfolio/src/main/webapp/script.js:104]: Might want to stick with the fetch API, it's more modern and I think the tutorials had you all using it.
https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -71,30 +71,40 @@ function fetchComments() {
 
 window.onload = (event) => {
   fetchComments();
-  toggleCommentBox();
+  changeCommentFormDisplay
```
Comment (augerm) [inline portfolio/src/main/webapp/script.js:74]: `changeCommentFormDisplay` --> `changeCommentFormDisplay()`

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (swalsh15) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:86]: An exception is thrown if verification fails - I'll add a comment to make that clear.

**Pull request merged**